node/src/chain/mod.rs

```diff
@@ -157,24 +165,44 @@ impl ChainSrv {
             &self.most_recent_block.header,
             &self.eligible_provisioners,
             db,
+            vm,
         );
 
         anyhow::Ok(0)
     }
 
-    async fn accept_block<DB: database::DB>(
+    async fn accept_block<DB: database::DB, VM: vm::VMExecution>(
         &mut self,
         db: &Arc<RwLock<DB>>,
+        vm: &Arc<RwLock<VM>>,
         blk: &Block,
     ) -> anyhow::Result<()> {
+        // Verify Block Header
+        self.verify_block_header(
+            db,
+            &self.most_recent_block.header,
+            &blk.header,
+        )
+        .await?;
+
         // Reset Consensus
-        self.upper.abort();
+        self.upper.abort().await;
+
+        // Persist block in consistency with the VM state update
+        {
+            let vm = vm.read().await;
+            db.read().await.update(|t| {
+                t.store_block(blk, true)?;
 
-        // TODO: Ensure block is valid
-        // TODO: Call ExecuteStateTransition
+                // Accept block transactions into the VM
+                if blk.header.cert.step == 3 {
+                    return vm.finalize(blk);
```

Comment:
Why we finalize and return? Shouldn't we continue with deleting from mempool and updating the most_recent_block?

Reply:
This is called within the scope of `update` closure. If finalize/accept returns Result::Ok, we continue with `most_recent_block`. 

If they return an Err:
- we rollback the stored block
- we return from accept_block without updating any state.

Reply:
You're right
I missed the scope, my bad